test/ntttcp_test.py

```diff
@@ -0,0 +1,470 @@
+class TestNtttcp:
+    expected_throughput = 70  # need evaluate
+    loopback_interface = "127.0.0.1"
+    set_duration_time = 5
+
+    # todo need add setup and teardown
+    def run_test(self, receiver_cmd: str, sender_cmd: str) -> NtttcpResult:
+        print(f"Receiver command: {receiver_cmd}")
+        print(f"Sender command: {sender_cmd}")
+        #every receiver command is with "-D", so the receiver command can return
+        #at once. Otherwise, we need to use "subprocess.Popen"
+        with open("rec_result.txt", "wb") as rec_out:
+            subprocess.run([receiver_cmd], shell=True, stdout=rec_out)
+        rec_open = open("rec_result.txt", "r")
+        rec_out = rec_open.read()
+        rec_open.close()
+        time.sleep(1)
+        with open("send_result.txt", "wb") as send_out:
+            subprocess.run([sender_cmd], shell=True, stdout=send_out)
+        send_open = open("send_result.txt", "r")
+        send_out = send_open.read()
+        send_open.close()
+
+        return NtttcpResult(
+                rec_out,
+                send_out,
+                )
+
+    def combine_command(
+        self,
+        common_option: Optional[str] = "",
+        receiver_option: Optional[str] = "",
+        sender_option: Optional[str] = "",
+    ) -> Tuple[str, str]:
+        rec_cmd = f"ulimit -n 40960 && ntttcp -r{self.loopback_interface} -t {self.set_duration_time} -Q -D"
+        send_cmd = f"ulimit -n 40960 && ntttcp -s{self.loopback_interface} -t {self.set_duration_time} -Q"
+        if common_option:
+            rec_cmd = f"{rec_cmd} {common_option}"
+            send_cmd = f"{send_cmd} {common_option}"
+        if receiver_option:
+            rec_cmd = f"{rec_cmd} {receiver_option}"
+        if sender_option:
+            send_cmd = f"{send_cmd} {sender_option}"
+        return rec_cmd, send_cmd
+
+    def teardown(self):
+        subprocess.run("killall ntttcp", shell=True)
+        print("kill all ntttcp process to clean up")
+
+    def test_daemon(self) -> None:
+        common_option = "-V"
+        rec_cmd, send_cmd = self.combine_command(common_option=common_option)
+        result = self.run_test(rec_cmd, send_cmd)
+        parse_result = NtttcpResult(result.rec_stdout, result.send_stdout)
+        assert parse_result.is_daemon_running() is True
+        throughput = parse_result.get_throughput()
+        print("throughput =", throughput)
+
+    def test_multi_clients_mode(self) -> None:
+        receiver_option = "-M -V"
+        sender_option = "-L"
+        rec_cmd, send_cmd = self.combine_command(
+            receiver_option=receiver_option, sender_option=sender_option
+        )
+        result = self.run_test(rec_cmd, send_cmd)
+        parse_result = NtttcpResult(result.rec_stdout, result.send_stdout)
+        assert parse_result.is_multi_clients_mode() is True
+        throughput = parse_result.get_throughput()
+        print("throughput =", throughput)
+
+    def test_multi_ports_threads(self) -> None:
```

Comment:
Can we split this test cases into 2: one for TCP and another one for UDP. For sure, the two new functions should be be able to share some common logic which can be put in a shared function.

Reply:
done